File: src/net/chroma/math/geometry/Triangle.java
package net.chroma.math.geometry;

import net.chroma.math.Constants;
import net.chroma.math.ImmutableArrayMatrix3x3;
import net.chroma.math.ImmutableVector3;
import net.chroma.math.Vector3;
import net.chroma.math.raytracing.Ray;

/**
 * Chroma uses the right-hand-coordinate system. Think about three vertices specified counterclockwise on the floor.
 * The normal will always point upwards!
 * @author steinerb
 */
public class Triangle implements Geometry {

    private static final double EPSILON = 0.000001;

    private final ImmutableVector3 p0, p1, p2;
    private final ImmutableVector3 n;

    public Triangle(ImmutableVector3 p0, ImmutableVector3 p1, ImmutableVector3 p2, ImmutableVector3 n) {
        this.p0 = p0;
        this.p1 = p1;
        this.p2 = p2;
        this.n = n;

    }

    public Triangle(ImmutableVector3 p0, ImmutableVector3 p1, ImmutableVector3 p2) {
        this.p0 = p0;
        this.p1 = p1;
        this.p2 = p2;
        this.n = e1().crossProduct(e2()).normalize();
    }

    @Override
    public float intersect(Ray ray){

        ImmutableVector3 P, Q, T;
        float det;

        ImmutableVector3 E1 = e1();
        ImmutableVector3 E2 = e2();

        /* begin calculating determinant - also used to calculate U parameter */
        P = ray.getDirection().crossProduct(E2);

        det = E1.dot(P);

        /* if determinant is near zero, ray lies in plane of triangle */
        if(det > -Constants.DBL_EPSILON && det < Constants.DBL_EPSILON){
            return 0.f;
        }

        // backface culling?
//      #ifdef BACKFACECULLING
//    /* calculate distance from vert0 to ray origin */
//        T = ray->origin - p0;
//
//    /* calculate U parameter and test bounds */
//        u = T * P;
//        if(u < 0.0f || u > det)
//            return 0;
//
//    /* prepare to test V parameter */
//        Q = T % E1;
//
//    /* calculate V parameter and test bounds */
//        v = ray->direction * Q;
//        if(v < 0.0f || u + v > det)
//            return 0;
//
//        det = 1./det;
//        u *= det;
//        v *= det;
//    /* calculate t, scale parameters, ray intersects triangle */
//        return E2 * Q * det;
//        #else

        float invDet = 1.0f / det;

	    /* calculate distance from vert0 to ray origin */
        T = ray.getOrigin().subtract(p0);

	    /* calculate U parameter and test bounds */
        float u = T.dot(P) * invDet;
        if (u < 0.0f || u > 1.0f) {
            return 0.f;
        }

	    /* prepare to test V parameter */
        Q = T.crossProduct(E1);

	    /* calculate V parameter and test bounds */
        float v = ray.getDirection().dot(Q) * invDet;
        if (v < 0.0f || u + v > 1.0f) {
            return 0.f;
        }

	    /* calculate t, ray intersects triangle */
        float distance = E2.dot(Q) * invDet;

//        if(ray.getTMin() > distance || ray.getTMax() < distance) {
//            return 0.f;
//        }

        return distance ;
    }

    @Override
    public Geometry transpose(Vector3 transpose) {
        return new Triangle(p0.plus(transpose), p1.plus(transpose), p2.plus(transpose));
    }

    @Override
    public Geometry rotate(ImmutableArrayMatrix3x3 rotationY) {
        return new Triangle(rotationY.mult(p0), rotationY.mult(p1), rotationY.mult(p2));
    }

    @Override
    public ImmutableVector3 getNormal(ImmutableVector3 hitpoint) {
        return n;
    }

    @Override
    public final boolean isPlane() {
        return true;
    }

    private ImmutableVector3 e2() {
        return new ImmutableVector3(p2).subtract(p0);
    }

    private ImmutableVector3 e1() {
        return new ImmutableVector3(p1).subtract(p0);
    }
}

File: chroma-scala-core/src/main/scala/net/chromarenderer/math/geom/Triangle.scala
package net.chromarenderer.math.geom

import net.chromarenderer.math.{Constants, Vector3}
import net.chromarenderer.math.raytracing.Ray

/**
  * Chroma uses the right-hand-coordinate system. Think about three vertices specified counterclockwise on the floor.
  * The normal will always point upwards!
  *
  * @author bensteinert
  */

class Triangle(p0_in: Vector3, p1_in: Vector3, p2_in: Vector3, n_in: Vector3) {

  var p0: Vector3 = p0_in
  var p1: Vector3 = p1_in
  var p2: Vector3 = p2_in
  var n: Vector3 = n_in


  def this(p0: Vector3, p1: Vector3, p2: Vector3) {
    this(p0, p1, p2, (p1 - p0 % p2 - p0).norm())
  }

  def e1: Vector3 = {
    p1 - p0
  }

  def e2: Vector3 = {
    p2 - p0
  }


  def subdivide: Array[Triangle] = {

    val pOnE1 = p0 + p1 * 0.5f
    val pOnE2 = p0 + p2 * 0.5f
    val pOnE3 = p1 + p2 * 0.5f

    Array(
      new Triangle(p0, pOnE1, pOnE2, n),
      new Triangle(pOnE1, p1, pOnE3, n),
      new Triangle(pOnE3, p2, pOnE2, n),
      new Triangle(pOnE1, pOnE3, pOnE2, n)
    )
  }

  def intersect(ray: Ray): Float = {

    // Ignore intersection from the 'back' (backface culling)
    val dotProduct: Float = ray.dir * n
    if (dotProduct > 0) {
      return 0
    }

    val P = ray.dir % e2
    val det = e1 * P

    /* if determinant is near zero, ray lies in plane of triangle */
    if (det > -Constants.FLT_EPSILON && det < Constants.FLT_EPSILON) {
      return 0
    }

    val invDet: Float = 1 / det
    val T: Vector3 = ray.origin - p0
    /* calculate U parameter and test texture coord bounds */
    val u: Float = (T * P) * invDet
    if (u < 0 || u > 1) {
      return 0
    }

    val Q = T % e1
    /* calculate V parameter and test bounds */
    val v = (ray.dir * Q) * invDet
    if (v < 0.0f || u + v > 1.0f) {
      return 0
    }

    /* calculate distance, ray intersects triangle */
    (e2 * Q) * invDet
  }

}
